refactor: rebuild lsp command by api

glepnir opened the pull request:
cc @justinmk 

Fix #2068

ranjithshegde commented:
@glepnir #1984 already includes some of these changes

ii14 commented on plugin/lspconfig.lua:
```diff
@@ -0,0 +1,58 @@
+  end, require('lspconfig.util').get_managed_clients())
+end
+
+if vim.fn.exists 'g:lspconfig' == 1 then
+  return
+end
+vim.g.lspconfig = 1
```
Can you move it to the top? So defining `g:lspconfig` will disable it entirely.

ii14 commented on plugin/lspconfig.lua:
```diff
@@ -0,0 +1,58 @@
+local api = vim.api
+local lspconfig = require 'lspconfig'
```
Can `require 'lspconfig'` be inlined in all places where `lspconfig` is used? Then we don't force the module to be loaded.

glepnir replied:
done

ii14 commented on plugin/lspconfig.lua:
```diff
@@ -0,0 +1,63 @@
+end, {
+  nargs = '?',
+  complete = function(arg)
+    local lspconfig = require 'lspconfig'
+    local list = lspconfig.available_servers()
+    return vim.tbl_filter(function(s)
+      return string.match(s, '^' .. arg)
+    end, list)
```
```suggestion
    return vim.tbl_filter(function(s)
      return s:sub(1, #arg) == arg
    end, require 'lspconfig'.available_servers())
```

ii14 commented on plugin/lspconfig.lua:
```diff
@@ -0,0 +1,63 @@
+})
+
+api.nvim_create_user_command('LspStart', function(args)
+  local lspconfig = require 'lspconfig'
+  lspconfig.lsp_start(args.args)
```
```suggestion
  require 'lspconfig'.lsp_start(args.args)
```

ii14 replied:
@glepnir you've missed this one

glepnir replied:
oh sry. fix

ii14 commented on plugin/lspconfig.lua:
```diff
@@ -0,0 +1,63 @@
+})
+
+api.nvim_create_user_command('LspStop', function(cmd_args)
+  local lspconfig = require 'lspconfig'
+  for _, client in ipairs(lspconfig.util.get_clients_from_cmd_args(cmd_args)) do
+    client.stop()
+  end
```
```suggestion
  for _, client in ipairs(require 'lspconfig'.util.get_clients_from_cmd_args(cmd_args)) do
    client.stop()
  end
```

ii14 commented on plugin/lspconfig.lua:
```diff
@@ -0,0 +1,63 @@
+})
+
+api.nvim_create_user_command('LspRestart', function(cmd_args)
+  local lspconfig = require 'lspconfig'
+  lspconfig.lsp_restart(cmd_args)
```
```suggestion
  require 'lspconfig'.lsp_restart(cmd_args)
```

glepnir replied:
done

justinmk commented:
let's revisit after https://github.com/neovim/nvim-lspconfig/pull/1984 ? Though maybe this isn't needed if `nvim_create_user_command` serves the purpose well enough? 

What's the use case that isn't covered by `nvim_create_user_command` ?

glepnir commented:
This is different from #1984. I just replaced lspconfig.vim. and  fixed #2068  now we can use `LspInfo border=single` to set the border . why not provide a config to set the border. because I think the `lspconfig` just provide the server config. compare with `vim.cmd` use string format to generate the `autocmd` use api will be more better:) and  neovim use lua file in everywhere now :)

justinmk reviewed (approved):
yeah, this looks like good changes.

ranjithshegde commented:
This conflicts with #1984 because `lspconfig.lua` logic introduced in that PR is incompatible. All these `user_commands` are moved to `plugin/lspconfig.lua`, with `lua/lspconfig.lua` just creatin/assigning a metatable for lsp_configurations.

I would recommend please wait till #1984 is merged. then this PR is simply modifying the `border` key and make it accessible by `fargs/nargs`

glepnir commented:
@ranjithshegde  because you don't reply my review so I continue work on this . maybe this will be merge or not @ii14  what do you think?

ranjithshegde commented:
> @ranjithshegde because you don't reply my review so I continue work on this . maybe this will be merge or not @ii14 what do you think?

My apologies. For whatever reason I wasn't subscribed to this thread. 

If you could please wait until that is merged, then the job of this PR will be very very simple. All the logic will already have been introduced in #1984,
 then you will simply have to make the `border` key modifiable by user.

glepnir commented:
> now we can use LspInfo border=single to set the border . why not provide a config to set the border. because I think the lspconfig just provide the server config.

this project only provide the server config. not provide any other config should be better. so pass the argument into the cmdline is my way. not use a `border` key config .

ii14 commented:
> what do you think?

I don't know, I'm only reviewing the code :)
